File: CSharp/AspNetCore/CoreWebApp/Controllers/PadesSignatureServerKeyController.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;
using System.Threading.Tasks;
using CoreWebApp.Classes;
using CoreWebApp.Models;
using Lacuna.RestPki.Api;
using Lacuna.RestPki.Api.PadesSignature;
using Lacuna.RestPki.Client;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Options;

namespace CoreWebApp.Controllers {

    [Route("api/[controller]")]
    public class PadesSignatureServerKeyController : Controller {

        private IHostingEnvironment hostingEnvironment;
        private RestPkiConfig restPkiConfig;

        public PadesSignatureServerKeyController(IHostingEnvironment hostingEnvironment, IOptions<RestPkiConfig> optionsAcessor) {
            this.hostingEnvironment = hostingEnvironment;
            this.restPkiConfig = optionsAcessor.Value;
        }

        [HttpPost]
        public async Task<SignatureCompleteResponse> Post(string userfile) {

            var storage = new Storage(hostingEnvironment);
            var client = Util.GetRestPkiClient(restPkiConfig);

            // Read the certificate from a PKCS#12 file.
            //var cert = Util.GetSampleCertificateFromPKCS12(storage);

            // Alternative option: Get the certificate from Microsoft CryptoAPI.
            var cert = Util.GetSampleCertificateFromMSCAPI();

            // Get an instance of the PadesSignatureStarter class, responsible for receiving the signature elements and start the
            // signature process
            var signatureStarter = new PadesSignatureStarter(client) {

                // Set the unit of measurement used to edit the pdf marks and visual representations
                MeasurementUnits = PadesMeasurementUnits.Centimeters,

                // Set the signature policy.
                SignaturePolicyId = StandardPadesSignaturePolicies.PkiBrazil.BasicWithPkiBrazilCerts,

                // For this sample, we'll use the Lacuna Test PKI as our security context in order to accept our test certificate used
                // above ("Pierre de Fermat"). This security context should be used ***** FOR DEVELOPMENT PUPOSES ONLY *****
                SecurityContextId = new Guid("803517ad-3bbc-4169-b085-60053a8f6dbf"),


                // Set a visual representation for the signature
                VisualRepresentation = new PadesVisualRepresentation() {

                    // The tags {{name}} and {{br_cpf_formatted}} will be substituted according to the user's certificate
                    //
                    //		name             : full name of the signer
                    //		br_cpf_formatted : if the certificate is ICP-Brasil, contains the signer's CPF
                    //
                    // For a full list of the supported tags, see: https://github.com/LacunaSoftware/RestPkiSamples/blob/master/PadesTags.md
                    Text = new PadesVisualText("Signed by {{name}} ({{br_cpf_formatted}})") {

                        // Specify that the signing time should also be rendered
                        IncludeSigningTime = true,

                        // Optionally set the horizontal alignment of the text ('Left' or 'Right'), if not set the default is Left
                        HorizontalAlign = PadesTextHorizontalAlign.Left

                    },

                    // We'll use as background the image in Content/PdfStamp.png
                    Image = new PadesVisualImage(storage.GetPdfStampContent(), "image/png") {

                        // Opacity is an integer from 0 to 100 (0 is completely transparent, 100 is completely opaque).
                        Opacity = 50,

                        // Align the image to the right
                        HorizontalAlign = PadesHorizontalAlign.Right

                    },

                    // Position of the visual representation. We have encapsulated this code in a method to include several
                    // possibilities depending on the argument passed. Experiment changing the argument to see different examples
                    // of signature positioning (valid values are 1-6). Once you decide which is best for your case, you can place
                    // the code directly here.
                    Position = PadesVisualElements.GetVisualPositioning(client, 1)
                }
            };

            // Set the signer certificate.
            signatureStarter.SetSignerCertificate(cert.RawData);

            // Below we'll either set the PDF file to be signed. Prefer passing a path or a stream instead of the file's contents
            // as a byte array to prevent memory allocation issues with large files.

            // If the "userfile" URL argument is set, it will contain the filename under the "App_Data" folder. Otherwise 
            // (signature with server file), we'll sign a sample document.
            if (string.IsNullOrEmpty(userfile)) {
                signatureStarter.SetPdfToSign(storage.GetSampleDocPath());
            } else {
                Stream userFileStream;
                if (!storage.TryOpenRead(userfile, out userFileStream)) {
                    throw new Exception("File not found");
                }
                signatureStarter.SetPdfToSign(userFileStream);
            }

            /*
				Optionally, add marks to the PDF before signing. These differ from the signature visual representation in that
				they are actually changes done to the document prior to signing, not binded to any signature. Therefore, any number
				of marks can be added, for instance one per page, whereas there can only be one visual representation per signature.
				However, since the marks are in reality changes to the PDF, they can only be added to documents which have no previous
				signatures, otherwise such signatures would be made invalid by the changes to the document (see property
				PadesSignatureStarter.BypassMarksIfSigned). This problem does not occurr with signature visual representations.

				We have encapsulated this code in a method to include several possibilities depending on the argument passed.
				Experiment changing the argument to see different examples of PDF marks (valid values are 1-3). Once you decide which 
                is best for your case, you can place the code directly here.
			*/
            //signatureStarter.PdfMarks.Add(PadesVisualElements.GetPdfMark(storage, 1));

            // Call the Start() method, which initiates the signature. This yields the parameters for the signature using the
            // certificate
            var signatureParams = signatureStarter.Start();

            // Get the key from a PKCS#12 file.
            var pkey = cert.GetRSAPrivateKey();

            // Alternative option: Get the key from Microsoft CryptoAPI.
            // TODO!

            // Perform the signature usign the parameters returned by Rest PKI with the signer's key.
            var signature = pkey.SignHash(signatureParams.ToSignHash, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);

            // Get an instance of the PadesSignatureFinisher2 class, responsible for completing the signature process
            var signatureFinisher = new PadesSignatureFinisher2(client) {

                // Set the token for this signature (acquired previously and passed back here by the angular controller)
                Token = signatureParams.Token,

                // Set the signature
                Signature = signature

            };

            // Call the FinishAsync() method, which finalizes the signature process and returns a SignatureResult object
            var signatureResult = await signatureFinisher.FinishAsync();

            // The "Certificate" property of the SignatureResult object contains information about the certificate used by the user
            // to sign the file.
            var signerCert = signatureResult.Certificate;

            // At this point, you'd typically store the signed PDF on a database or storage service. For demonstration purposes, we'll
            // store the PDF on our "storage mock", which in turn stores the PDF on the App_Data folder.

            // The SignatureResult object has various methods for writing the signature file to a stream (WriteToAsync()), local file (WriteToFileAsync()),
            // open a stream to read the content (OpenReadAsync()) and get its contents (GetContentAsync()). Avoid the method GetContentAsync() to prevent
            // memory allocation issues with large files.
            string filename;
            using (var signatureStream = await signatureResult.OpenReadAsync()) {
                filename = await storage.StoreAsync(signatureStream, ".pdf");
            }

            // Pass the following fields to be used on signature-results template:
            // - The signature filename, which can be used to provide a link to the file
            // - The user's certificate
            var response = new SignatureCompleteResponse() {
                Filename = filename,
                Certificate = new Models.CertificateModel(signerCert)
            };

            return response;
        }
    }
}

File: Java/sample-spring-mvc/src/main/java/sample/controller/PadesSignatureServerKeyController.java
package sample.controller;

import com.lacunasoftware.restpki.*;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import sample.Application;
import sample.util.PadesVisualElements;
import sample.util.Util;

import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.security.GeneralSecurityException;
import java.security.PrivateKey;
import java.security.Signature;
import java.security.cert.Certificate;
import java.util.UUID;

@Controller
public class PadesSignatureServerKeyController {

	/**
	 * This action performs a PAdES signature using REST PKI and a PKCS#12 certificate. It renders
	 * the signature page.
	 */
	@RequestMapping(value = "/pades-signature-server-key", method = {RequestMethod.GET})
	public String get(
			@RequestParam(value = "userfile", required = false) String userfile,
			Model model
	) throws IOException, RestException, GeneralSecurityException {

		// Read the certificate from a PKCS#12 file.
		Certificate certificate = Util.getSampleCertificateFromPKCS12();

		// Alternative option: Get the certificate from Microsoft CryptoAPI.
		//Certificate certificate = Util.getSampleCertificateFromMSCAPI();

		// Get an instance of the PadesSignatureStarter2 class, responsible for receiving the
		// signature elements and start the signature process.
		PadesSignatureStarter2 signatureStarter = new PadesSignatureStarter2(Util.getRestPkiClient());

		// Set the signer certificate.
		signatureStarter.setSignerCertificateRaw(certificate.getEncoded());

		// Set the unit of measurement used to edit the pdf marks and visual representations.
		signatureStarter.setMeasurementUnits(PadesMeasurementUnits.Centimeters);

		// Set the signature policy. For this sample, we'll use the Lacuna Test PKI in order to
		// accept our test certificate used above ("Pierre de Fermat"). This security context
		// should be used FOR DEVELOPMENT PUPOSES ONLY. In production, you'll typically want one of
		// the alternatives below. For more test certificates, see:
		// https://github.com/LacunaSoftware/RestPkiSamples/blob/master/TestCertificates.md
		signatureStarter.setSignaturePolicy(SignaturePolicy.PadesBasic);
		signatureStarter.setSecurityContext(SecurityContext.lacunaTest);

		// Create a visual representation for the signature.
		signatureStarter.setVisualRepresentation(PadesVisualElements.getVisualRepresentation());

		// Below we'll either set the PDF file to be signed. Prefer passing a path or a stream
		// instead of the file's contents as a byte array to prevent memory allocation issues with
		// large files.
		if (userfile != null && !userfile.isEmpty()) {

			// If the URL argument "userfile" is filled, it means the user was redirected here by
			// UploadController (signature with file uploaded by user). We'll set the path of the file
			// to be signed, which was saved in the temporary folder by UploadController (such a file
			// would normally come from your application's database).
			signatureStarter.setPdfToSign(Application.getTempFolderPath().resolve(userfile));

		} else {

			// If both userfile is null, this is the "signature with server file" case. We'll set the
			// file to be signed by passing its path.
			signatureStarter.setPdfToSign(Util.getSampleDocPath());

		}

		// Optionally, add marks to the PDF before signing. These differ from the signature visual
		// representation in that they are actually changes done to the document prior to signing,
		// not binded to any signature. Therefore, any number of marks can be added, for instance one
		// per page, whereas there can only be one visual representation per signature. However,
		// since the marks are in reality changes to the PDF, they can only be added to documents
		// which have no previous signatures, otherwise such signatures would be made invalid by the
		// changes to the document (see field bypassMarksIfSigned of PadesSignatureStarter2). This
		// problem does not occur with signature visual representations.
		//
		// We have encapsulated this code in a method to include several possibilities depending on
		// the argument passed. Experiment changing the argument to see different examples of PDF
		// marks (valid values are 1-4). Once you decide which is best for your case, you can place
		// the code directly here.
		//signatureStarter.addPdfMark(PadesVisualElements.getPdfMark(1));

		// Call the start() method, which initiates the signature. This yields the parameters for
		// the signature using the certificate.
		SignatureStartResult result = signatureStarter.start();

		// Get the key from a PKCS#12 file.
		PrivateKey pkey = (PrivateKey) Util.getSampleKeyFromPKCS12();

		// Alternative option: Get the key from Microsoft CryptoAPI.
		//PrivateKey pkey = (PrivateKey) Util.getSampleKeyFromMSCAPI();

		// Initialize  the signature using the parameters returned by Rest PKI with the signer's key.
		Signature signature = Signature.getInstance(result.getSignatureAlgorithm());
		signature.initSign(pkey);
		// Set the toSignData, provided by Rest PKI.
		byte[] toSignData = result.getToSignDataRaw();
		signature.update(toSignData, 0, toSignData.length);
		// Perform the signature.
		byte[] sig = signature.sign();

		//
		// Get an instance of the PadesSignatureFinisher2 class, responsible for completing the
		// signature process.
		PadesSignatureFinisher2 signatureFinisher = new PadesSignatureFinisher2(Util.getRestPkiClient());

		// Set the token for this signature, provided by Rest PKI.
		signatureFinisher.setToken(result.getToken());

		// Set the signature
		signatureFinisher.setSignature(sig);

		// Call the finish() method, which finalizes the signature process and returns a
		// SignatureResult object.
		SignatureResult signatureResult = signatureFinisher.finish();

		// The "certificate" field of the SignatureResult object contains information about the
		// certificate used by the user to sign the file.
		PKCertificate signerCert = signatureResult.getCertificate();

		// At this point, you'd typically store the signed PDF on your database. For demonstration
		// purposes, we'll store the PDF on a temporary folder and return to the page an identifier
		// that can be used to download it.

		String filename = UUID.randomUUID() + ".pdf";

		// The SignatureResult object has various methods for writing the signature file to a stream
		// (writeTo()), local file (writeToFile()), open a stream to read the content (openRead())
		// and get its contents (getContent()). For large files, avoid the method getContent() to
		// avoid memory allocation issues.
		signatureResult.writeToFile(Application.getTempFolderPath().resolve(filename));

		// Render the signature page (templates/pades-signature-info.html).
		model.addAttribute("signerCert", signerCert);
		model.addAttribute("filename", filename);
		return "pades-signature-server-key";
	}
}
